src/BSON/ArrayList.c

```diff
@@ -0,0 +1,390 @@
+/*
+ * Copyright 2014-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <php.h>
+#include <ext/standard/base64.h>
+#include <Zend/zend_interfaces.h>
+#include <Zend/zend_operators.h>
+#include <ext/standard/php_var.h>
+#include <zend_smart_str.h>
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "php_phongo.h"
+#include "phongo_error.h"
+#include "phongo_bson_encode.h"
+#include "BSON/ArrayList_arginfo.h"
+#include "BSON/Iterator.h"
+
+zend_class_entry* php_phongo_arraylist_ce;
+
+/* Initialize the object from a HashTable and return whether it was successful.
+ * An exception will be thrown on error. */
+static bool php_phongo_arraylist_init_from_hash(php_phongo_arraylist_t* intern, HashTable* props)
+{
+	zval* data;
+
+	if ((data = zend_hash_str_find(props, "data", sizeof("data") - 1)) && Z_TYPE_P(data) == IS_STRING) {
+		zend_string* decoded = php_base64_decode_str(Z_STR_P(data));
+
+		intern->bson = bson_new_from_data((const uint8_t*) ZSTR_VAL(decoded), ZSTR_LEN(decoded));
+		zend_string_free(decoded);
+
+		return intern->bson != NULL;
+	}
+
+	phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT, "%s initialization requires \"data\" string field", ZSTR_VAL(php_phongo_arraylist_ce->name));
+	return false;
+}
+
+static HashTable* php_phongo_arraylist_get_properties_hash(phongo_compat_object_handler_type* object, bool is_temp)
+{
+	php_phongo_arraylist_t* intern;
+	HashTable*              props;
+
+	intern = Z_OBJ_ARRAYLIST(PHONGO_COMPAT_GET_OBJ(object));
+
+	PHONGO_GET_PROPERTY_HASH_INIT_PROPS(is_temp, intern, props, 2);
+
+	if (!intern->bson) {
+		return props;
+	}
+
+	{
+		zval data, length;
+
+		ZVAL_STR(&data, php_base64_encode((const unsigned char*) bson_get_data(intern->bson), intern->bson->len));
+		zend_hash_str_update(props, "data", sizeof("data") - 1, &data);
+
+		ZVAL_LONG(&length, intern->bson->len);
+		zend_hash_str_update(props, "length", sizeof("length") - 1, &length);
```

Comment:
Note that `php_phongo_arraylist_get_properties_hash` is used for serialize output, so you're adding an extra field that is never read by `php_phongo_arraylist_init_from_hash`.

Having `php_phongo_arraylist_get_properties_hash` conditionally add a `length` property would be tricky as we internally cache the props with `PHONGO_GET_PROPERTY_HASH_INIT_PROPS`. I think you're better off adding length after the fact in `get_debug_info` or just omitting it entirely.

Reminder that `var_export` uses `get_properties`, so it wouldn't get the length in that case (nor should it IMO).

I'm inclined to say just omit the `length` field. That's what we do for Binary objects already.

Reply:
Removed the `length` field from properties and serialised output.